vortex-array/src/compute/take.rs

```diff
@@ -10,6 +11,25 @@ pub trait TakeFn {
 pub fn take(array: impl AsRef<Array>, indices: impl AsRef<Array>) -> VortexResult<Array> {
     let array = array.as_ref();
     let indices = indices.as_ref();
+
+    if !indices.dtype().is_int() || indices.dtype().is_nullable() {
+        vortex_bail!(
+            "Take indices must be a non-nullable integer type, got {}",
+            indices.dtype()
+        );
+    }
+
+    // If the indices are large enough, it's faster to canonicalize the indices and take those.
+    if indices.len() >= array.len() && PrimitiveArray::try_from(indices).is_err() {
```

Comment:
Are there any encodings that don't immediately convert the indices into a PrimitiveArray already? It seems a bit confusing to have this at the top-level but then have every encoding `into_primitive` anyway.

Reply:
Right now most of the time you flatten the indices but I think we shoulnd't do that. There's a couple of optimizations here to explore and we should first look at them before changing behaviour of top level function.